### PR TITLE
Bifrost providers and append bootstrap

### PR DESCRIPTION
**Pull request by AhmedSoliman**
Bifrost providers and append bootstrap

Bootstrapping providers and loglet traits and the `Bifrost::append()` (single record) path.

---
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/restatedev/restate/pull/1147).
* #1149
* __->__ #1147
* #1141

### CODE REVIEW

File: crates/bifrost/src/loglet.rs [hunk 1]
```diff
@@ -54,3 +67,68 @@ impl LogletOffset {
...
+    loglet: Arc<dyn Loglet>,
+}
+
+impl LogletWrapper {
+    pub fn new(base_lsn: Lsn, loglet: Arc<dyn Loglet>) -> Self {
+        Self { base_lsn, loglet }
+    }
+}
+
+/// A loglet represents a single log. Systems that support multiple logs
```
Comment (tillrohrmann) [inline crates/bifrost/src/loglet.rs:103]: The second sentence seems to be cut off.

File: crates/bifrost/src/loglet.rs [hunk 2]
```diff
@@ -54,3 +67,68 @@ impl LogletOffset {
...
+///
+/// Loglets are required to follow these rules:
+/// - Loglet implementations must be Send + Sync (internal mutability is required)
+/// - Loglets must strictly adhere to the consistency requirements as the interface calls
+///   that it, if an append returns an offset, it **must** be durably committed.
```
Comment (tillrohrmann) [inline crates/bifrost/src/loglet.rs:108]: "that it" seems not fitting.

File: crates/bifrost/src/bifrost.rs [hunk 3]
```diff
@@ -99,4 +146,216 @@ impl BifrostInner {
...
+            .await;
+        assert!(matches!(res, Err(Error::Shutdown)));
+        // Validate the watchdog has called the provider::start() function.
+        assert!(logs_contain("Starting in-memory loglet provider"));
+        assert!(logs_contain("Shutting down in-memory loglet provider"));
+        assert!(logs_contain("Bifrost watchdog shutdown complete"));
+        Ok(())
+    }
+
+    #[test(tokio::test(start_paused = true))]
```
Comment (tillrohrmann) [inline crates/bifrost/src/bifrost.rs:319]: I didn't know about `start_paused = true` and then using `tokio::time::Instant`. This is really cool :-)

File: crates/bifrost/src/bifrost.rs [hunk 4]
```diff
@@ -99,4 +146,216 @@ impl BifrostInner {
...
+    #[traced_test]
```
Comment (tillrohrmann) [inline crates/bifrost/src/bifrost.rs:320]: Why is `traced_test` necessary in this test?

**Review by tillrohrmann — APPROVED** [review]
Really nice work @AhmedSoliman 👏. The changes look good to me :-) +1 for merging.

File: crates/bifrost/src/bifrost.rs (hunk 4 above)
Comment (AhmedSoliman) [reply crates/bifrost/src/bifrost.rs:320]: Not necessary. Will remove. I'm just lazy, since it automatically sets up logging in tests :)

**Pull request merged**